### ISSUE TITLE
[Cluster] Error when creating a user after setting up SF as resident SP

### ISSUE THREAD

**Issue by pubudup** [wso2/product-is#2135]
Pre-conditions:
- Follow doc [1] and configure SF as SP in IS
- Follow the doc at [2] and configure SF as IDP

Steps:-
1. Created a user in management console
2. Save the user
3. Login to Salesforce and check if the user is created

Expected Results:-
- User should be created in Salesforce

Actual Results:-
- User is not created in Salesforce
- The following exception is triggered in IS

```
TID: [-1234] [] [2018-01-19 12:15:51,641]  INFO {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Provisioning separator is not defined, hence using default provisioning separator
TID: [-1234] [] [2018-01-19 12:15:52,947] ERROR {org.wso2.carbon.identity.provisioning.ProvisioningThread} -   Provisioning for Entity sfuser@qa.com For operation = POST
org.wso2.carbon.identity.provisioning.IdentityProvisioningException: Error in invoking provisioning operation for the user
	at org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector.createUser(SalesforceProvisioningConnector.java:261)
	at org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector.provision(SalesforceProvisioningConnector.java:105)
	at org.wso2.carbon.identity.provisioning.ProvisioningThread.call(ProvisioningThread.java:68)
	at org.wso2.carbon.identity.provisioning.ProvisioningThread.call(ProvisioningThread.java:32)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.http.client.ClientProtocolException
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:106)
	at org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector.createUser(SalesforceProvisioningConnector.java:234)
	... 7 more
Caused by: org.apache.http.ProtocolException: Target host is not specified
	at org.apache.http.impl.conn.DefaultRoutePlanner.determineRoute(DefaultRoutePlanner.java:69)
	at org.apache.http.impl.client.InternalHttpClient.determineRoute(InternalHttpClient.java:124)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:183)
	... 10 more

```

[1] https://docs.wso2.com/display/IS540/Logging+in+to+Salesforce+using+the+Identity+Server
[2] https://docs.wso2.com/display/IS540/Outbound+Provisioning+with+Salesforce#OutboundProvisioningwithSalesforce-ConfiguringSalesforce

**Comment by pubudup** [comment]
Pack: IAM 5.4.0 update4
Tags: 
Severity/Critical
Priority/High

**Comment by omindu** [comment]
@pubudup Make sure the Domain name in `Salesforce Provisioning Configuration` section is
in the format `https://<DOMAIN>-dev-ed.my.salesforce.com`. Note the `https://` prefix.

**Comment by pubudup** [comment]
I was able to get rid of this particular issue once the domain is provided. But now I am seeing the error message below.

```
ERROR {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Received response status code: 404 text: Not Found
```

**Comment by pubudup** [comment]
Salesforce DEBUG logs attached below.

```
TID: [-1234] [] [2018-01-22 07:13:17,769]  INFO {org.wso2.carbon.core.services.util.CarbonAuthenticationUtil} -  'admin@wso2.com@carbon.super [-1234]' logged in at [2018-01-22 07:13:17,769+0000]
TID: [-1234] [] [2018-01-22 07:13:44,801] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnectorFactory} -  Salesforce provisioning connector created successfully.
TID: [-1234] [] [2018-01-22 07:13:44,874] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: ProfileId , value is: 1516356091429
TID: [-1234] [] [2018-01-22 07:13:44,875] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Email , value is: pubudup@wso2.com
TID: [-1234] [] [2018-01-22 07:13:44,875] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LocaleSidKey , value is: en_US
TID: [-1234] [] [2018-01-22 07:13:44,875] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LanguageLocaleKey , value is: en_US
TID: [-1234] [] [2018-01-22 07:13:44,876] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsOfflineUser , value is: false
TID: [-1234] [] [2018-01-22 07:13:44,876] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: TimeZoneSidKey , value is: America/Los_Angeles
TID: [-1234] [] [2018-01-22 07:13:44,876] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Username , value is: primary_sfuser13.qa.com_carbon.super_salesforce.com@qa.com
TID: [-1234] [] [2018-01-22 07:13:44,876] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Alias , value is: PDP
TID: [-1234] [] [2018-01-22 07:13:44,877] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsCallCenterAutoLogin , value is: false
TID: [-1234] [] [2018-01-22 07:13:44,877] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LastName , value is: DP
TID: [-1234] [] [2018-01-22 07:13:44,877] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: EmailEncodingKey , value is: ISO-8859-1
TID: [-1234] [] [2018-01-22 07:13:44,877] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsMarketingUser , value is: false
TID: [-1234] [] [2018-01-22 07:13:44,878] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  JSON object of User
{
  "ProfileId": "1516356091429",
  "Email": "pubudup@wso2.com",
  "LocaleSidKey": "en_US",
  "LanguageLocaleKey": "en_US",
  "UserPermissionsOfflineUser": false,
  "TimeZoneSidKey": "America/Los_Angeles",
  "Username": "primary_sfuser13.qa.com_carbon.super_salesforce.com@qa.com",
  "Alias": "PDP",
  "UserPermissionsCallCenterAutoLogin": false,
  "LastName": "DP",
  "EmailEncodingKey": "ISO-8859-1",
  "UserPermissionsMarketingUser": false
}
TID: [-1234] [] [2018-01-22 07:13:44,883] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Built user endpoint url : https://pubudup-dev-ed.my.salesforce.com/services/data/41.0/sobjects/user/
TID: [-1234] [] [2018-01-22 07:13:47,163] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authentication to salesforce returned with response code: 200
TID: [-1234] [] [2018-01-22 07:13:47,166] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authenticate response: {
  "access_token": "00D7F000003g61h!AQMAQN_7pUmmi1sO0zSmgWA8mfcYGCe3rQ2.9yyXgrcYNvPfCyU.wWCCsgrBlMCLsNhzXe_VWBt1W3uN7_2Xd9AAz6DbJZuz",
  "signature": "fQI5C+gvrWYa0VEUW6F2IQ4DG7RsYSbulvwYDT5H2Fo=",
  "instance_url": "https://pubudup-dev-ed.my.salesforce.com",
  "id": "https://login.salesforce.com/id/00D7F000003g61hUAA/0057F000000oWycQAE",
  "token_type": "Bearer",
  "issued_at": "1516605226992"
}
TID: [-1234] [] [2018-01-22 07:13:47,166] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Access token is: 00D7F000003g61h!AQMAQN_7pUmmi1sO0zSmgWA8mfcYGCe3rQ2.9yyXgrcYNvPfCyU.wWCCsgrBlMCLsNhzXe_VWBt1W3uN7_2Xd9AAz6DbJZuz
TID: [-1234] [] [2018-01-22 07:13:47,168] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Setting authorization header for method: POST as follows,
TID: [-1234] [] [2018-01-22 07:13:47,169] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authorization: OAuth 00D7F000003g61h!AQMAQN_7pUmmi1sO0zSmgWA8mfcYGCe3rQ2.9yyXgrcYNvPfCyU.wWCCsgrBlMCLsNhzXe_VWBt1W3uN7_2Xd9AAz6DbJZuz
TID: [-1234] [] [2018-01-22 07:13:48,314] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  HTTP status 404 creating user
TID: [-1234] [] [2018-01-22 07:13:48,315] ERROR {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Received response status code: 404 text: Not Found
TID: [-1234] [] [2018-01-22 07:13:48,316] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Error response : {"ProfileId":"1516356091429","Email":"pubudup@wso2.com","LocaleSidKey":"en_US","LanguageLocaleKey":"en_US","UserPermissionsOfflineUser":false,"TimeZoneSidKey":"America/Los_Angeles","Username":"primary_sfuser13.qa.com_carbon.super_salesforce.com@qa com","Alias":"PDP","UserPermissionsCallCenterAutoLogin":false,"LastName":"DP","EmailEncodingKey":"ISO-8859-1","UserPermissionsMarketingUser":false}
TID: [-1234] [] [2018-01-22 07:13:48,317] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Returning created user's ID: null

```

**Comment by pubudup** [comment]
Increasing the priority to Blocker since the scenario cannot be completed due to this issue.

**Comment by pubudup** [comment]
![sfp1](https://user-images.githubusercontent.com/13449310/35215902-c3f421e8-ff8b-11e7-8887-47b78e85fbe6.jpg)

**Comment by pubudup** [comment]
Corrected version as v41.0. Now I'm getting a 400 as response status.

```
TID: [-1234] [] [2018-01-22 11:45:07,515] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnectorFactory} -  Salesforce provisioning connector created successfully.
TID: [-1234] [] [2018-01-22 11:45:07,540] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: ProfileId , value is: 1516356091429
TID: [-1234] [] [2018-01-22 11:45:07,541] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Email , value is: pubudup@wso2.com
TID: [-1234] [] [2018-01-22 11:45:07,542] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LocaleSidKey , value is: en_US
TID: [-1234] [] [2018-01-22 11:45:07,542] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LanguageLocaleKey , value is: en_US
TID: [-1234] [] [2018-01-22 11:45:07,542] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsOfflineUser , value is: false
TID: [-1234] [] [2018-01-22 11:45:07,543] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: TimeZoneSidKey , value is: America/Los_Angeles
TID: [-1234] [] [2018-01-22 11:45:07,543] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Username , value is: xxx@qa.com
TID: [-1234] [] [2018-01-22 11:45:07,543] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: Alias , value is: PDP
TID: [-1234] [] [2018-01-22 11:45:07,544] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsCallCenterAutoLogin , value is: false
TID: [-1234] [] [2018-01-22 11:45:07,544] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: LastName , value is: DP
TID: [-1234] [] [2018-01-22 11:45:07,544] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: EmailEncodingKey , value is: ISO-8859-1
TID: [-1234] [] [2018-01-22 11:45:07,545] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  The key is: UserPermissionsMarketingUser , value is: false
TID: [-1234] [] [2018-01-22 11:45:07,545] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  JSON object of User
{
  "ProfileId": "1516356091429",
  "Email": "pubudup@wso2.com",
  "LocaleSidKey": "en_US",
  "LanguageLocaleKey": "en_US",
  "UserPermissionsOfflineUser": false,
  "TimeZoneSidKey": "America/Los_Angeles",
  "Username": "xxx@qa.com",
  "Alias": "PDP",
  "UserPermissionsCallCenterAutoLogin": false,
  "LastName": "DP",
  "EmailEncodingKey": "ISO-8859-1",
  "UserPermissionsMarketingUser": false
}
TID: [-1234] [] [2018-01-22 11:45:07,546] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Built user endpoint url : https://pubudup-dev-ed.my.salesforce.com/services/data/v41.0/sobjects/user/
TID: [-1234] [] [2018-01-22 11:45:08,516] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authentication to salesforce returned with response code: 200
TID: [-1234] [] [2018-01-22 11:45:08,517] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authenticate response: {
  "access_token": "xxxxxxxxxxxxxxxxxxx",
  "signature": "5qRLMpSKbEAAdMELN4o0lFyeTURi2i4EHZfpmzpBvro=",
  "instance_url": "https://pubudup-dev-ed.my.salesforce.com",
  "id": "https://login.salesforce.com/id/00D7F000003g61hUAA/0057F000000oWycQAE",
  "token_type": "Bearer",
  "issued_at": "1516621508433"
}
TID: [-1234] [] [2018-01-22 11:45:08,518] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Access token is: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
TID: [-1234] [] [2018-01-22 11:45:08,518] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Setting authorization header for method: POST as follows,
TID: [-1234] [] [2018-01-22 11:45:08,519] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Authorization: OAuth 00D7F000003g61h!AQMAQN_7pUmmi1sO0zSmgWA8mfcYGCe3rQ2.9yyXgrcYNvPfCyU.wWCCsgrBlMCLsNhzXe_VWBt1W3uN7_2Xd9AAz6DbJZuz
TID: [-1234] [] [2018-01-22 11:45:09,351] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  HTTP status 400 creating user
TID: [-1234] [] [2018-01-22 11:45:09,351] ERROR {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Received response status code: 400 text: Bad Request
TID: [-1234] [] [2018-01-22 11:45:09,351] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Error response : {"ProfileId":"1516356091429","Email":"pubudup@wso2.com","LocaleSidKey":"en_US","LanguageLocaleKey":"en_US","UserPermissionsOfflineUser":false,"TimeZoneSidKey":"America/Los_Angeles","Username":"xxx@qa.com","Alias":"PDP","UserPermissionsCallCenterAu oLogin":false,"LastName":"DP","EmailEncodingKey":"ISO-8859-1","UserPermissionsMarketingUser":false}
TID: [-1234] [] [2018-01-22 11:45:09,352] DEBUG {org.wso2.carbon.identity.provisioning.connector.salesforce.SalesforceProvisioningConnector} -  Returning created user's ID: null

```

**Comment by pubudup** [comment]
Looks like the profile id provided was incorrect. I have fixed the issue now. Closing the ticket.
